veps/sig-compute/node-pressure-eviction.md

```diff
@@ -0,0 +1,229 @@
+
+## Update/Rollback Compatibility
+
+Attempting to enable the NodePressureEvictionLiveMigration feature gate on previous versions of KubeVirt will fail when this feature gate does not exist.
+
+## Functional Testing Approach
+
+This feature can be tested by modifying the kubelet’s config on a set of nodes. Below are some example values to use. These example values set the soft eviction memory threshold high in order to make it easier to invoke the node pressure eviction. The evictionMaxPodGracePeriod corresponds to how long the pod can block shutting down until it is force killed. This value translates to how long the system has to evacuate the VMI via live migration before the pod dies.
+
+```yaml
+evictionSoft:
+  memory.available: "3500Mi"
+evictionSoftGracePeriod:
+  memory.available: "5s"
+evictionMaxPodGracePeriod: 300
+evictionHard:
+  memory.available: "10Mi"
+```
+
+To test invoking node pressure eviction, fill up a node with VMI’s that are configured with EvictionStrategy: LiveMigrate until the eviction memory threshold is hit. Then verify the migration is triggered and succeeds before the source pod is killed.
+
+## Implementation Phases
+
+Phase 0 - extend the EvictionStrategy to work with Node Pressure Eviction
+Phase 1 - consider live migration time optimizations to narrow the time window in which live migrations are active during Node Pressure Eviction.
+
+## Feature Lifecycle Phases
+
+### Alpha
+
+The initial release of this feature will be feature gated and considered an Alpha. What we're interested in here is gaining additional experience with what kubelet level tunings regarding node pressure provide the best results, and gaining a better understanding of usage patterns.
+
+### Beta
+
+Once we feel comfortable that this feature delivers on the goal of providing best effort live migrations across the majority of VM workloads, we'll graduate this feature to Beta.
+
+### GA
+
+Graduation to GA will occur once we have a clear signal that this feature behaves as expected in production and we have a clear set of recommendations for how different kubelet configs impact different types of VM workloads.
+
+# Open Questions
+
+## Recording Cause of Migration on VMIM
+
+Today when a VMIM is created, we don’t have context into why the system created the VMIM. It would be nice to have information tied to the VMIM that indicates why the migration was invoked… such as “API Eviction”, “Node Pressure Eviction, “virt-launcher update”, etc…
+
+## Migration Time Optimizations
+
+Even with a substantial grace period, there is some risk that the live migration will not complete in time before the VMI pod is killed when Node Pressure Eviction takes place. This is because we don’t have a mechanism to block the pod from being killed after the evictionMaxPodGracePeriod is exceeded or a hard eviction threshold is hit.
```

Comment:
> This is because we don’t have a mechanism to block the pod from being killed after the evictionMaxPodGracePeriod is exceeded or a hard eviction threshold is hit.

I wonder if an NRI plugin can do that.

Reply:
>I wonder if an NRI plugin can do that.

let's say that is possible. would it be reasonable for us to block node pressure eviction? It seems like this would be circumventing node pressure eviction in a way that side steps the contract defined in the kubelet's config.

Reply:
Yeah I agree. I've glanced at Kubernetes' eviction manager's code again and the call to evict pods is blocking, so until an eviction is complete kubelet will never evict another pod.

We can still consider NRIs to inject some eviction metadata though.